### PR TITLE
Add --log-level flags to ocitools and runtimetest

### PR DESCRIPTION
**Pull request by wking**
Details in the commit messages.

### CODE REVIEW

File: cmd/runtimetest/main.go [hunk 1]
```diff
@@ -263,3 +271,21 @@ func main() {
 		}
 	}
 }
+
+func main() {
+	app := cli.NewApp()
+	app.Name = "runtimetest"
+	app.Version = "0.0.1"
+	app.Usage = "Compare the environment with an OCI configuration"
+	//app.UsageText = "runtimetest [options]"  <-- in urfave/cli
```
Comment (liangchenye) [inline cmd/runtimetest/main.go]: Remove this?

Comment (wking) [inline cmd/runtimetest/main.go]: On Mon, Jun 06, 2016 at 02:04:54AM -0700, 梁辰晔 (Liang Chenye) wrote:
> > +func main() {
> > +	app := cli.NewApp()
> > +	app.Name = "runtimetest"
> > +	app.Version = "0.0.1"
> > +	app.Usage = "Compare the environment with an OCI configuration"
> > +	//app.UsageText = "runtimetest [options]"  <-- in urfave/cli
> 
> Remove this?

If you like, but I wanted to remember when #96 gets fixed.

Comment (mrunalp) [inline cmd/runtimetest/main.go]: Yeah, no harm in keeping this as a reminder.

**Comment by mrunalp** [comment]
LGTM

**Comment by mrunalp** [comment]
Upon testing I realized that we also need a --debug to test_runtime.sh to enable this.

**Comment by wking** [comment]
On Mon, Jun 06, 2016 at 04:53:05PM -0700, Mrunal Patel wrote:
> Upon testing I realized that we also need a --debug to
> test_runtime.sh to enable this.

Possibly added with 477e29f, but I don't have enough of this installed
as root to test it out ;).  It seemed to work when I made a few
uncomitted tweaks to work around that.

File: cmd/runtimetest/main.go (hunk 1 above)
Comment (liangchenye) [inline cmd/runtimetest/main.go]: >> If you like, but I wanted to remember when #96 gets fixed.
OK

**Comment by liangchenye** [comment]
I tested 477e29f, it works.
But runtimetest seems to recognize only one argument other than two: `/runtimetest--log-level=debug`.
Debugging on it...

**Comment by wking** [comment]
On Mon, Jun 06, 2016 at 07:47:02PM -0700, 梁辰晔 (Liang Chenye) wrote:
> I tested 477e29f, it works.  But runtimetest seems to recognize only
> one argument other than two: `/runtimetest--log-level=debug`.

I'm not sure what's going on there.  Here's tweaks for my root-less
test:

  $ git describe --always
  477e29f
  $ git diff
  diff --git a/test_runtime.sh b/test_runtime.sh
  index 71fa87d..18b7164 100755
  --- a/test_runtime.sh
  +++ b/test_runtime.sh
  @@ -61,8 +61,7 @@ if ! command -v ${RUNTIME} > /dev/null; then
          error "Runtime ${RUNTIME} not found in the path"
   fi

  -mkdir -p /testroot
  -TMPDIR=$(mktemp -p /testroot -d)
  +TMPDIR=$(mktemp -d)
   TESTDIR=${TMPDIR}/busybox
   mkdir -p ${TESTDIR}

  @@ -79,7 +78,7 @@ tar -xf  rootfs.tar.gz -C ${TESTDIR}
   cp runtimetest ${TESTDIR}

   pushd $TESTDIR > /dev/null
  -ocitools generate "${TEST_ARGS[@]}" --rootfs ""
  +python -c 'import sys; print(sys.argv)' ocitools generate "${TEST_ARGS[@]}" --rootfs ""
   popd > /dev/null

   TESTCMD="${RUNTIME} start $(uuidgen)"

And without -d:

  $ ./test_runtime.sh
  -----------------------------------------------------------------------------------
  VALIDATING RUNTIME: runc
  -----------------------------------------------------------------------------------
  ['-c', 'ocitools', 'generate', '--args', '/runtimetest', '--rootfs', '']
  open /run/runc/c7e8bc8c-993c-482b-8268-b6d35c5ec43d/state.json: permission denied
  Runtime runc failed validation

And with -d:

  $ ./test_runtime.sh -d debug
  -----------------------------------------------------------------------------------
  VALIDATING RUNTIME: runc
  -----------------------------------------------------------------------------------
  ['-c', 'ocitools', 'generate', '--args', '/runtimetest', '--args', '--log-level=debug', '--rootfs', '']
  open /run/runc/a210a181-b380-46c9-bed8-00efb3fe905e/state.json: permission denied
  Runtime runc failed validation

In both cases, we have --args setup the way ocitools-generate(1)
suggests (--args $ARG0 --args $ARG1 …).

**Comment by wking** [comment]
On Mon, Jun 06, 2016 at 08:55:49PM -0700, W. Trevor King wrote:
>   ['-c', 'ocitools', 'generate', '--args', '/runtimetest', '--args', '--log-level=debug', '--rootfs', '']

And running this directly gives the expected config:

  $ ./ocitools generate --args /runtimetest --args --log-level=debug --rootfs ''
  $ jq .process.args config.json 
  [
    "/runtimetest",
    "--log-level=debug"
  ]

**Comment by liangchenye** [comment]
@wking I found it is a bug of [cmd/runtimetest/main.go](
https://github.com/opencontainers/ocitools/blob/7f267cdc0766c26af6ad3ca760477791eaf5f43a/cmd/runtimetest/main.go#L79)
Args are separated by '\0' in /proc/1/cmdline, when using `string()`, the '\0' will be dropped.
I'll submit a patch.

**Comment by mrunalp** [comment]
Needs rebase.

**Comment by wking** [comment]
On Tue, Jun 07, 2016 at 02:25:02PM -0700, Mrunal Patel wrote:
> Needs rebase. 

I'll wait for #104, since fixing that is more important, and this will
probably need another rebase after that lands ;).

**Comment by wking** [comment]
On Tue, Jun 07, 2016 at 02:26:26PM -0700, W. Trevor King wrote:
> Tue, Jun 07, 2016 at 02:25:02PM -0700, Mrunal Patel:
> > Needs rebase. 
> 
> I'll wait for #104…

Rebased around #104 with 477e29f → c2beb36.

File: cmd/runtimetest/main.go [hunk 2]
```diff
@@ -291,7 +292,14 @@ func validateMountsExist(spec *rspec.Spec) error {
 	return nil
 }
 
-func main() {
+func validate(context *cli.Context) {
```
Comment (mrunalp) [inline cmd/runtimetest/main.go]: Needs to return error like #104

File: test_runtime.sh [hunk 3]
```diff
@@ -3,11 +3,23 @@ set -o errexit
...
+	exit 1
+fi
+
 RUNTIME="runc"
+TEST_ARGS=('--args' '/runtimetest')
 KEEP=0 # Track whether we keep the test directory around or clean it up
 
 usage() {
-	echo "$0 -r <runtime> -k -h"
+	echo "$0 -d <log-level> -r <runtime> -k -h"
```
Comment (mrunalp) [inline test_runtime.sh]: WDYT of replacing -d by -l ?

**Comment by mrunalp** [comment]
Also you missed "validating mounts exist" which is still visible in the output.

File: test_runtime.sh (hunk 3 above)
Comment (wking) [inline test_runtime.sh]: On Wed, Jun 08, 2016 at 03:42:57PM -0700, Mrunal Patel wrote:
> >  KEEP=0 # Track whether we keep the test directory around or clean it up
> >  
> >  usage() {
> > -	echo "$0 -r <runtime> -k -h"
> > +	echo "$0 -d <log-level> -r <runtime> -k -h"
> 
> WDYT of replacing -d by -l ?

Done in c2beb36 → 2c377ad.

File: cmd/runtimetest/main.go (hunk 2 above)
Comment (wking) [inline cmd/runtimetest/main.go]: On Wed, Jun 08, 2016 at 03:42:11PM -0700, Mrunal Patel wrote:
> > -func main() {
> > +func validate(context *cli.Context) {
> 
> Needs to return error like #104 

Fixed in c2beb36 → 2c377ad.

**Comment by wking** [comment]
On Wed, Jun 08, 2016 at 03:46:40PM -0700, Mrunal Patel wrote:
> Also you missed "validating mounts exist" which is still visible in
> the output.

Grr, sloppy rebase.  Fixed in 2c377ad → 70fd17b.

**Comment by mrunalp** [comment]
LGTM. Thanks!

**Pull request merged**